### PR TITLE
Add support to private_vpc (ie: with no public address)

### LINKED ISSUE #14: Support launching clusters into private VPCs
Some users work in environments where they want to or have to launch clusters into VPCs with no public subnets.

This seems to be [a fairly common use-case](https://issues.apache.org/jira/issues/?jql=project%20%3D%20SPARK%20AND%20component%20%3D%20EC2%20AND%20text%20~%20%22private%20vpc%22%20ORDER%20BY%20updated%20DESC), so I think we  should support it if it does not add too much complexity.

A design goal for this feature should be to automatically do the appropriate setup whether the subnet we are launching into is public or private. That means, if possible, the user shouldn't have to specify anything.

I _think_ this is possible because we know what VPC we're launching into (either the user's default VPC, or an explicitly specified one), and we know what subnet we're launching into (either the VPC's default subnet, or an explicitly specified one). Flintrock should be able to query AWS for information about the subnet and figure out whether to use public or private addresses automatically.

### PR DESCRIPTION
**Pull request by luhhujbb**
This PR makes the following changes:
* support of private_vpc (ie: with public ip address like this : https://github.com/chaordic/flintrock/pull/4/files)
* add client-source option since in private mode, the IP returned by checkip.amazonaws.com will be a nonsense in most of cases (using a VPN, for instance) and it's sounds good to have all good ports configured automatically in this case, even if custom security groups can be added. 
* client-source support both cidr address or security group.

Default behaviour remains identical except that it logs when flintrock detect private mode instead of raising an exception.
Fixes #14

### CODE REVIEW

**Comment by nchammas** [comment]
Hey @luhhujbb, thanks for submitting this PR. I will take a look at it over the holidays!

File: flintrock/ec2.py
Comment (nchammas) [inline flintrock/ec2.py]: It seems like if we have a new property to explicitly detect if we're running in a private subnet, `self.private_net`, we should use it here too (and just below) to switch between public and private addresses.

File: flintrock/ec2.py
Comment (luhhujbb) [reply flintrock/ec2.py]: * client-source doesn't add arbitrary rules, a user in a private environment doesn't want to think about different ports which are already defined in flintrock. One only wants basic rules to be setup where the flintrock client run. 
* client-source also disable public ip security rules configuration which is on a private environment quite useful in terms of security.

File: flintrock/ec2.py
Comment (luhhujbb) [reply flintrock/ec2.py]: Maybe I'm wrong, but a @property decorator implies that the the following code will be executed each times the property is get, that is not efficient. This way it reduces aws api call.

**Comment by luhhujbb** [comment]
Test passes with USE_AWS_CREDENTIALS=true in the following condition:
* public subnet (default behaviour)
* private subnet, with client-source option configured with my vpn IP.

File: flintrock/ec2.py
Comment (luhhujbb) [reply flintrock/ec2.py]: Fixed.

File: flintrock/ec2.py
Comment (nchammas) [inline flintrock/ec2.py]: I don't think a double underscore variable name is appropriate here. Those are for special methods and "class-private" members, which is not the case here.

File: flintrock/ec2.py
Comment (nchammas) [inline flintrock/ec2.py]: My intention in recommending a `@property` was that all the logic would be contained within it. Yes, there is a penalty in that the query to AWS will be made (ignoring any caching that boto3 may do) every time you access the property, but that minor downside is outweighed in my view by the clearer API. The property doesn't get called that often.

So I'd say to either put everything in the property method, or go back to having it just be a plain old property like you did before. It's not a big deal either way.

File: tests/conftest.py
Comment (nchammas) [inline tests/conftest.py]: Did you need to make this change to get tests to pass? I'd keep it out of the PR since it's unrelated to the topic at hand.

File: flintrock/flintrock.py
Comment (nchammas) [inline flintrock/flintrock.py]: I think we need to improve a few things here.

1. I'd rename this option to `--ec2-authorize-access-from`. It's clearer, and it categorizes the option under EC2 where it belongs.
2. I'd also add a brief help text explaining what this option does and what input it takes, since it may not be obvious to everyone. Since this option disables the automatic detection of the client IP, we should call that out too.
3. We should somehow validate the input here since we are expecting either a CIDR or a security group ID. A user might, as I did, assume they can provide a plain IP address, which then leads to the following error:

    ```
    botocore.exceptions.ClientError: An error occurred (InvalidParameterValue) when calling the 
    AuthorizeSecurityGroupIngress operation: CIDR block 172.31.38.197 is malformed
    ```

    Either that, or we should accept a plain IP address and automatically covert it into a `/32` CIDR block.

    Take a look at `--ec2-tag`  for an example of how to validate user input.
4. Since we are adding a new option, we need to add it to the [config template](https://github.com/nchammas/flintrock/blob/master/flintrock/config.yaml.template).

**Comment by nchammas** [comment]
This is looking good @luhhujbb. I will test this out tomorrow and try to get a private VPC working in my environment. I think I need to setup a [NAT gateway](https://docs.aws.amazon.com/vpc/latest/userguide/vpc-nat-gateway.html) to enable the private VPC to still talk to the outside world.

Do you mind if I push some commits to your branch? It will mostly be to tweak some method names or message text that is sent to the user. I've been very slow with the review feedback, so for small things I'd like to spare you the time and just do them myself.

If you're OK with that, let me know and just make sure that the "Allow edits from maintainers" checkbox is enabled in the right side-bar.

**Comment by luhhujbb** [comment]
Hi, @nchammas, of course you can edit, the box "Allow edits from maintainers" is checked.

**Comment by nchammas** [comment]
I've successfully created a private subnet that accesses the Internet via a NAT gateway on a public subnet in the same VPC. I've also successfully tested out this PR by launching a cluster in that environment. So far so good. 👍

I've pushed the latest changes from master onto your fork, @luhhujbb, and my next steps will be to retest your PR given the recent changes introduced by #285. I will also push some edits as noted in my previous comment.

(By the way, for the future you may want to make contributions from a branch on your fork, as opposed to from master. It will make pulling updates from the upstream source easier down the line.)

**Comment by kedar700** [comment]
Any update on when this might get merged

**Comment by nchammas** [comment]
Hello @kedar700. Sorry, I haven't been able to look at this PR recently.

I think this PR has the right implementation but, before it can be merged in, it needs some tweaks and needs merge conflicts to be resolved.

It's my fault for leaving this PR in limbo for so long. (Sorry @luhhujbb!)

**Comment by luhhujbb** [comment]
I've rebase from master. The python 3.5 compatibility was difficult to handle. Maybe it should be dropped since python 3.5 reaches its end of life.
 
`authorize-access-from` is now an array of configurable security-group/network cidr, it allows flexible configuration in a private environment.

**Comment by nchammas** [comment]
Thank you for updating this PR @luhhujbb. I've dropped Python 3.5 support in #329 and will take a look at your updates here later this week. Hopefully we can get this merged in this time around!

**Comment by luhhujbb** [comment]
@nchammas I've just rebase. Waiting for your review. Thanks for your time.

File: flintrock/core.py
Comment (nchammas) [inline flintrock/core.py:80]: Nit: Could we rename this `private_network()`?

File: flintrock/ec2.py
Comment (nchammas) [inline flintrock/ec2.py:85]: I'd delete this comment. It's unnecessary.

File: flintrock/flintrock.py
Comment (nchammas) [inline flintrock/flintrock.py:332]: I forget if we already discussed this, but does the `add-slaves` command automatically respect this setting? I assume yes, since this is done at the level of the cluster security group, but just confirming.

**Review by nchammas — COMMENTED** [review]
Took a quick look now. Looks good. Will take a closer look and retest later this week.

File: flintrock/ec2.py
Comment (luhhujbb) [reply flintrock/ec2.py:85]: Ok

File: flintrock/flintrock.py
Comment (luhhujbb) [reply flintrock/flintrock.py:332]: Slaves get their security groups from cluster master ones, so it's not needed:

https://github.com/nchammas/flintrock/blob/d3f769e40d1cce742996412b6a2dc2956870fc93/flintrock/ec2.py#L258-L260

**Comment by nchammas** [comment]
@luhhujbb - I made some changes to the PR to tweak the wording or style in various places, and also to refactor how we are validating the CLI input. I also think I fixed some bugs, like `.appends -> .append`. I tested my changes by launching a cluster into a private VPC with both Hadoop and Spark installed.

1. What do you think of the latest changes?
2. Can you test and confirm that the current state of the PR still works for your private VPC use case?

**Review by nchammas — APPROVED** [review]

**Comment by luhhujbb** [comment]
@nchammas First of all many thanks for your time !
1. I've check all your commits and it looks good.
2. I've successfully launch a production cluster using this version of flintrock. Everythinks is Ok. Using multipe "ec2-autorize-access-from" allows us to meet all our production use cases.

**Comment by nchammas** [comment]
Thanks for persisting through the lengthy review/refactor process! I'm glad we got this in.

**Pull request merged**
